### PR TITLE
improve the elasticsearch module

### PR DESCRIPTION
**Pull request by Avilir**
This PR improve the local elasticsearch module by preventing it to leave leftovers if deployment failed.
it also :
  * setting the storageclass as optional parameter
  * Delete the namespace if it is already exists, and try to continue deployment
  * Delete the project if PVC did not create

Signed-off-by: Avi Layani <alayani@redhat.com>

### CODE REVIEW

File: tests/libtest/test_elasticsearch.py [hunk 1]
```diff
@@ -2,35 +2,49 @@
...
 """
+# Internal modules
 import logging
 import time
 
+# 3rd party modules
 from elasticsearch import Elasticsearch, exceptions as esexp
 
-from ocs_ci.ocs import defaults
+# Local modules
```
Comment (keesturam) [inline tests/libtest/test_elasticsearch.py:12]: to be removed?

File: tests/libtest/test_elasticsearch.py [hunk 2]
```diff
@@ -113,14 +127,30 @@ def smallfile_run(self, es):
...
+            * deploy the elasticsearch module
+            * connect to it
+            * run a simple SmallFile benchmark (to verify usability)
+            * dump the results to a file
+            * push the results from the file to the Dev. ES.
+            * teardown the environment
 
         """
 
+        log.info("Test with 'Dummy' Storageclass")
```
Comment (keesturam) [inline tests/libtest/test_elasticsearch.py:141]: why is this needed?

File: tests/libtest/test_elasticsearch.py [hunk 3]
```diff
@@ -113,14 +127,30 @@ def smallfile_run(self, es):
...
+            self.es = ElasticSearch(sc="dummy")
+        except Exception:
+            log.info("Raised as expected !")
+
+        log.info("Test with 'Real' Storageclass")
+        try:
+            self.es = ElasticSearch()
+        except Exception as ex:
+            log.error("Raise as expected !")
+            raise ex
```
Comment (ramkiperiy) [inline tests/libtest/test_elasticsearch.py:152]: @Avilir Can we add specific exception here https://github.com/red-hat-storage/ocs-ci/blob/master/ocs_ci/ocs/exceptions.py which will be more specific. Vasu mentioned that this might cause failure with Python3.8

Based on his comment I updated the code as in this line https://github.com/red-hat-storage/ocs-ci/blob/master/tests/e2e/scale/test_scale_12_OCS_worker_nodes_and_6000_PVCs.py#L97

File: tests/libtest/test_elasticsearch.py (hunk 1 above)
Comment (Avilir) [reply tests/libtest/test_elasticsearch.py:12]: no, just reordering the imports.

File: tests/libtest/test_elasticsearch.py (hunk 2 above)
Comment (Avilir) [reply tests/libtest/test_elasticsearch.py:141]: to verify that the cleanup incase of pvc not bounded is working, and for that i am using "not exist (dummy)" storageclass.

File: tests/libtest/test_elasticsearch.py (hunk 3 above)
Comment (Avilir) [reply tests/libtest/test_elasticsearch.py:152]: i will update it.

Comment (ramkiperiy) [reply tests/libtest/test_elasticsearch.py:152]: This comment applies to line no 154 & 168 in ocs_ci/ocs/elasticsearch.py file too.

File: ocs_ci/ocs/elasticsearch.py [hunk 4]
```diff
@@ -308,22 +334,35 @@ def get_password(self):
...
-        self.ocp.delete(yaml_file=self.dumper_file)
-        log.info("Deleting the es resource")
-        self.ocp.delete(yaml_file=self.crd)
+        try:
+            log.info("Deleting the dumper client pod")
+            self.ocp.delete(yaml_file=self.dumper_file)
+        except CommandFailed:
+            # in case of the es-dumper did not deployed yet, trying to delete it
+            # will failed.
+            pass
```
Comment (ramkiperiy) [inline ocs_ci/ocs/elasticsearch.py:348]: Good to add a log.warn message that expected pod is not deployed so skipped deleting.

File: ocs_ci/ocs/elasticsearch.py [hunk 5]
```diff
@@ -251,14 +265,21 @@ def _deploy_es(self):
...
+            self.pvc_obj = create_pvc(
+                sc_name=self.args.get("sc") or constants.CEPHBLOCKPOOL_SC,
+                namespace=self.namespace,
+                pvc_name="elasticsearch-data-quickstart-es-default-0",
+                access_mode=constants.ACCESS_MODE_RWO,
+                size="10Gi",
+            )
+
+            # Make sure the PVC bound, or delete it and raise exception
+            wait_for_resource_state(self.pvc_obj, constants.STATUS_BOUND)
```
Comment (ramkiperiy) [inline ocs_ci/ocs/elasticsearch.py:278]: Do we have the validation if this PVC exists or not, sorry if I missed here in the coding somewhere. also would suggest to add in cleanup function too.

Comment (Avilir) [reply ocs_ci/ocs/elasticsearch.py:278]: before creation of the PVC, there is a validation that the namespace is not exist (or delete it if needed) at lines 134-143, so this will make sure the PVC is not exist at this point.

**Review by ramkiperiy — APPROVED** [review]
LGTM

**Review by OdedViner — APPROVED** [review]
